### PR TITLE
Fix #2: Added hooks for the hidden and visible algorithms

### LINKED ISSUE #2: Hooks for "the document becomes visible/hidden"
[[
The specification has "now visible algorithm" and "now hidden algorithm" but misses extension points at the end of those algorithms. It would be great to add such extension points so other specifications could cleanly define behaviour when the document visibility state changes.
]]
http://lists.w3.org/Archives/Public/public-web-perf/2014Sep/0034.html

### PR DESCRIPTION
**Pull request by plehegar**
Fixing https://github.com/w3c/page-visibility/issues/2

### CODE REVIEW

**Comment by plehegar** [comment]
the idea for those hooks is simply to allow other specs to add their additional steps based on the hidden state. raf doesn't react to a change of hidden state per say. Instead, raf gets triggered if the hidden is false. We'll need to work the quirks of raf separately imho without touching page visibility.

**Comment by natduca** [comment]
+1

File: index.html [hunk 1]
```diff
@@ -452,8 +452,12 @@ <h3 id="sec-processing-model"><span class="secno">4.4 </span>Processing Model</h
 			</p>
 		</li>
 </ol>
+<p>At certain points in the algorithms below it is said that a
+<dfn id="document-is-visible">document is visible</dfn> or a <dfn id="document-is-hidden">document is hidden</dfn>. These are
+hooks for <a href="#other-applicable-specifications">other applicable specifications</a> to add additional steps depending on the state of the <a href="#dom-document-hidden"><code>hidden</code></a> attribute.
+</p>
```
Comment (marcoscaceres) [inline index.html]: This is redundant. Having these hooks here doesn't help. Hooks need to be to the actual definitions.

File: index.html [hunk 2]
```diff
@@ -452,8 +452,12 @@ <h3 id="sec-processing-model"><span class="secno">4.4 </span>Processing Model</h
 			</p>
 		</li>
 </ol>
+<p>At certain points in the algorithms below it is said that a
+<dfn id="document-is-visible">document is visible</dfn> or a <dfn id="document-is-hidden">document is hidden</dfn>. These are
+hooks for <a href="#other-applicable-specifications">other applicable specifications</a> to add additional steps depending on the state of the <a href="#dom-document-hidden"><code>hidden</code></a> attribute.
+</p>
 <p>
-        The <b><dfn id="now-visible-algorithm">now visible algorithm</dfn></b> runs the following steps synchronously:
+  The <b><dfn id="now-visible-algorithm">now visible algorithm</dfn></b> runs the following steps synchronously:
```
Comment (marcoscaceres) [inline index.html]: Nit: you don't need the `b` elements.

File: index.html [hunk 3]
```diff
@@ -466,6 +470,9 @@ <h3 id="sec-processing-model"><span class="secno">4.4 </span>Processing Model</h
                 <p>Fire a simple event named <code>visibilitychange</code> that bubbles, isn't cancelable, and has no default action,
                 at the <a href="http://www.w3.org/TR/html5/dom.html#document">Document</a>.</p>
             </li>
+            <li>
+              <p>Run <a href="#document-is-visible">document is visible</a>.</p>
```
Comment (marcoscaceres) [inline index.html]: You can't run, "document is visible" as it's a concept, not an algorithm. Maybe there needs to be a `[[visible]] slot` that is set at various points. So, a document is **visible** if the slot is set to true. A document can only be set to visible if X conditions are met.

File: index.html [hunk 4]
```diff
@@ -486,6 +493,9 @@ <h3 id="sec-processing-model"><span class="secno">4.4 </span>Processing Model</h
         <p>Fire a simple event named <code>visibilitychange</code> that bubbles, isn't cancelable, and has no default action,
         at the <a href="http://www.w3.org/TR/html5/dom.html#document">Document</a>.</p>
     </li>
+    <li>
+      <p>Run <a href="#document-is-hidden">document is hidden</a>.</p>
```
Comment (marcoscaceres) [inline index.html]: As above.

File: index.html [hunk 5]
```diff
@@ -507,6 +517,16 @@ <h3 id="sec-vendor-prefix"><span class="secno">4.5 </span>Vendor Prefixes</h3>
...
+<p>When vendor-neutral extensions to this specification are needed,
+either this specification can be updated accordingly, or an
+extension specification can be written that overrides the
+requirements in this specification. When someone applying this
+specification to their activities decides that they will recognize
+the requirements of such an extension specification, it becomes an
+<dfn id="other-applicable-specifications" title="other applicable specifications">applicable
+specification</dfn> for the purposes of conformance requirements in
+this specification.
+</p>
```
Comment (marcoscaceres) [inline index.html:488]: Please remove this. This encourages :monkey: patching. Specially: 

> extension specification can be written that overrides the requirements in this specification.

Extensions to this specification should be done in _this_ specification - as part of a errata or living standards model.

File: index.html (hunk 3 above)
Comment (plehegar) [inline index.html]: How will this slot differ from the actual definition of Document.hidden?

**Comment by plehegar** [comment]
I'll need to better understand the use cases here before I can make another proposal. The visibility state is given by Document.hidden. The change of state is given by the visibilitychange event. What is needed for other specifications?

File: index.html (hunk 3 above)
Comment (mounirlamouri) [inline index.html]: What about something like:
Extension point: specifications that want to react when a <dfn>document became visible</dfn> can run their own algorithms here.

The &lt;dfn&gt; gives something to link to and it pushes specifications to not monkey patch but just use this as an event.
@marcoscaceres and @plehegar what do you think?

File: index.html (hunk 4 above)
Comment (mounirlamouri) [inline index.html]: ditto

**Comment by igrigorik** [comment]
@mounirlamouri @plehegar @marcoscaceres rereading the discussion here, what about the ~following:

http://w3c.github.io/page-visibility/#sec-processing-model
- move step 3 from "now visible" and "now hidden" algorithms out into a substep following both
- prior (or after?) to above step add an extension point as suggested in https://github.com/w3c/page-visibility/pull/4#discussion_r24290704. 

Would that be sufficient?

**Comment by plehegar** [comment]
@igrigorik I believe that would be sufficient.

**Comment by igrigorik** [comment]
@plehegar @marcoscaceres @mounirlamouri PTAL: https://rawgit.com/w3c/page-visibility/plh/hooks/index.html#sec-processing-model - how does that look?

**Comment by plehegar** [comment]
ok for me. @marcoscaceres ?

**Comment by marcoscaceres** [comment]
This still doesn't do what we want (apologies for not being clear). We need to name algorithms. I also spotted a whole bunch of other markup and cross ref issues. I guess I need to stop complaining and just fix stuff to do want... my comments about monkey patching also remain.

**Comment by marcoscaceres** [comment]
I'll do a once over now... I suggest we close this.

**Comment by igrigorik** [comment]
@marcoscaceres I want to make sure we do resolve this; I'll wait for your pull before closing this? :)

**Comment by marcoscaceres** [comment]
@igrigorik https://github.com/w3c/page-visibility/pull/5 ... :feelsgood:.

**Comment by igrigorik** [comment]
:+1:

**Pull request closed without merging**